### PR TITLE
turn on concurrent GC on Linux

### PR DESCRIPTION
**Pull request by leirocks**
make it build with concurrent GC turned on
background zeroing enabled with a replacement of SLIST (implemented with critical section, may search for better implementation)
enable write barrier in large block
make it able to turn off hardware writebarrier on windows to simulate Linux
on Linux all write barrier pages are assuming dirty while scanning, to make the tests pass. need to remove this after the annotation is done

### CODE REVIEW

**Comment by leirocks** [comment]
@dotnet-bot test Linux tests please

**Comment by leirocks** [comment]
@dotnet-bot test Linux tests please

File: lib/Common/CommonBasic.h [hunk 1]
```diff
@@ -8,6 +8,16 @@
 #include "CommonDefines.h"
 #define _CRT_RAND_S         // Enable rand_s in the CRT
 
+#pragma warning(push)
+#if defined(PROFILE_RECYCLER_ALLOC) || defined(HEAP_TRACK_ALLOC) || defined(ENABLE_DEBUG_CONFIG_OPTIONS)
+#ifdef __clang__
+namespace std { struct type_info; }
+typedef std::type_info type_info;
```
Comment (jianchun) [inline lib/Common/CommonBasic.h]: Is "std::type_info" declared in "<typeinfo>"? If so may "using std::type_info" after "#include<typeinfo>".

**Comment by leirocks** [comment]
@dotnet-bot test Linux tests please

**Comment by leirocks** [comment]
@dotnet-bot test Linux tests please

**Comment by leirocks** [comment]
@dotnet-bot test Linux tests please

**Comment by leirocks** [comment]
@dotnet-bot test Linux tests please

**Comment by leirocks** [comment]
@dotnet-bot test Linux tests please

**Comment by leirocks** [comment]
@dotnet-bot test Ubuntu daily_ubuntu_linux_debug please

File: lib/Runtime/Library/BoundFunction.cpp [hunk 2]
```diff
@@ -189,6 +189,9 @@ namespace Js
                 // it is possible that the bound arguments are not marshalled yet.
                 for (uint i = 0; i < boundFunction->count; i++)
                 {
+                    //warning C6386: Buffer overrun while writing to 'newValues':  the writable size is 'boundFunction->count+argCount*8' bytes, but '40' bytes might be written.
+                    // there's throw with args.Info.Count == 0, so here won't hit buffer overrun, and __analyze_assume(argCount>0) does not work
+#pragma warning(suppress: 6386)
                     newValues[index++] = CrossSite::MarshalVar(scriptContext, boundFunction->boundArgs[i]);
```
Comment (jianchun) [inline lib/Runtime/Library/BoundFunction.cpp]: Strange for this seemingly unrelated prefast change here. Why master didn't hit it? Should this be done in master?

File: lib/Common/CommonDefines.h [hunk 3]
```diff
@@ -120,6 +120,21 @@
...
 
+#define BUCKETIZE_MEDIUM_ALLOCATIONS 1              // *** TODO: Won't build if disabled currently
+#define SMALLBLOCK_MEDIUM_ALLOC 1                   // *** TODO: Won't build if disabled currently
+#define LARGEHEAPBLOCK_ENCODING 1                   // Large heap block metadata encoding
+#define RECYCLER_WRITE_BARRIER                      // Write Barrier support
+#define IDLE_DECOMMIT_ENABLED 1                     // Idle Decommit
+#define RECYCLER_PAGE_HEAP                          // PageHeap support
+
+#ifdef RECYCLER_WRITE_BARRIER
+#if defined(__clang__)
```
Comment (jianchun) [inline lib/Common/CommonDefines.h]: >__clang__ [](start = 12, length = 9)

I think some `__clang__` uses should be `_WIN32`. We can use clang on Windows with hardware write barrier.

File: lib/Common/ConfigFlagsList.h [hunk 4]
```diff
@@ -1477,6 +1487,9 @@ FLAGR(Number, JITServerIdleTimeout, "Idle timeout in seconds to do the cleanup i
 FLAGR(Number, JITServerMaxInactivePageAllocatorCount, "Max inactive page allocators to keep before schedule a cleanup", 10)
 
 FLAGNR(Boolean, StrictWriteBarrierCheck, "Check write barrier setting on none write barrier pages", false)
+FLAGR (Boolean, WriteBarrierTest, "Always return true while checking barrier to test recycler regardless of annotation", DEFAULT_CONFIG_WriteBarrierTest)
+FLAGR (Boolean, EnableHardwareWriteBarrier, "Use to turn off write watch to test software write barrier on windows", DEFAULT_CONFIG_EnableHardwareWriteBarrier)
+FLAGR (Boolean, EnableBGFreeZero, "Use to turn off background freeing and zeroing to simulate linux", DEFAULT_CONFIG_EnableBGFreeZero)
```
Comment (jianchun) [inline lib/Common/ConfigFlagsList.h]: Should these flags be added to retail? Looks they should stay in debug/test builds?

Comment (leirocks) [inline lib/Common/ConfigFlagsList.h]: FLAGR means available in release, no?
if you want to test in free build I think you need to change the default value and rebuild. in free build the unnecessary code is optimized. 

---
In reply to: [87881989](https://github.com/Microsoft/ChakraCore/pull/1968#discussion_r87881989) [](ancestors = 87881989)

File: lib/Common/CommonDefines.h [hunk 5]
```diff
@@ -131,33 +146,27 @@
 #define ENABLE_BACKGROUND_PAGE_ZEROING 1
 #define ENABLE_BACKGROUND_PAGE_FREEING 1
 #define ENABLE_RECYCLER_TYPE_TRACKING 1
+#define SUPPORT_SLIST 1
```
Comment (leirocks) [inline lib/Common/CommonDefines.h]: here SLIST means the nonblocking implementation of interlocked singly linked list  in win32 , let me change it to SUPPORT_WIN32_SLIST

---
In reply to: [87880167](https://github.com/Microsoft/ChakraCore/pull/1968#discussion_r87880167) [](ancestors = 87880167)

**Comment by leirocks** [comment]
@dotnet-bot test Linux tests please

**Comment by leirocks** [comment]
@dotnet-bot test Ubuntu daily_ubuntu_linux_test please

File: lib/Common/ConfigFlagsList.h [hunk 6]
```diff
@@ -1477,6 +1487,9 @@ FLAGR(Number, JITServerIdleTimeout, "Idle timeout in seconds to do the cleanup i
 FLAGR(Number, JITServerMaxInactivePageAllocatorCount, "Max inactive page allocators to keep before schedule a cleanup", 10)
 
 FLAGNR(Boolean, StrictWriteBarrierCheck, "Check write barrier setting on none write barrier pages", false)
+FLAGNR(Boolean, WriteBarrierTest, "Always return true while checking barrier to test recycler regardless of annotation", DEFAULT_CONFIG_WriteBarrierTest)
+FLAGNR(Boolean, EnableHardwareWriteBarrier, "Use to turn off write watch to test software write barrier on windows", DEFAULT_CONFIG_EnableHardwareWriteBarrier)
```
Comment (digitalinfinity) [inline lib/Common/ConfigFlagsList.h]: which is another way of asking, under what circumstances will we use hardware write watch if GLOBAL_FORCE_USE_WRITE_BARRIER is set to true?

---
In reply to: [87929125](https://github.com/Microsoft/ChakraCore/pull/1968#discussion_r87929125) [](ancestors = 87929125)

Comment (leirocks) [inline lib/Common/ConfigFlagsList.h]: test the behavior on windows, without rebuilding, after the SWB work is done I think we can remove it.

---
In reply to: [87929333](https://github.com/Microsoft/ChakraCore/pull/1968#discussion_r87929333) [](ancestors = 87929333,87929125)

Comment (leirocks) [inline lib/Common/ConfigFlagsList.h]: just think again, looks this only useful now - while I assuming all write barrier pages are dirty. after annotation is done, we can't really test it that way. let me remove this.

---
In reply to: [87930847](https://github.com/Microsoft/ChakraCore/pull/1968#discussion_r87930847) [](ancestors = 87930847,87929333,87929125)

Comment (leirocks) [inline lib/Common/ConfigFlagsList.h]: looks my codeflow can't post message to the github

"my original idea was this, with GLOBAL_FORCE_USE_WRITE_BARRIER is on on windows,  we can change with this flag to test with either software write barrier or hardware write barrier. if we found a issue in software write barrier, we can use the same binary to verify if this issue can repro with hardware write barrier. do you think this is useful? if yes I suggest we keep this until the feature is done."

Comment (digitalinfinity) [inline lib/Common/ConfigFlagsList.h]: ah i see- so GLOBAL_FORCE_USE_WRITE_BARRIER does not actually force using software write barrier? it only allows you to use software write barrier everywhere? seems confusing- although given the explanation, I see the value of EnableHardwareWriteBarrier, but I would change some name (maybe GLOBAL_FORCE_USE_WRITE_BARRIER) so this isn't as confusing

---
In reply to: [87932027](https://github.com/Microsoft/ChakraCore/pull/1968#discussion_r87932027) [](ancestors = 87932027,87931407,87930847,87929333,87929125)

Comment (leirocks) [inline lib/Common/ConfigFlagsList.h]: let me change GLOBAL_FORCE_USE_WRITE_BARRIER to GLOBAL_ENABLE_WRITE_BARRIER for compile time enabling the WriteBarrierPtr wrapper
and change EnableHardwareWriteBarrier to ForceSoftwareWriteBarrier for runtime forcibly use software write barrier even on windows 

---
In reply to: [87938342](https://github.com/Microsoft/ChakraCore/pull/1968#discussion_r87938342) [](ancestors = 87938342,87932027,87931407,87930847,87929333,87929125)

**Comment by leirocks** [comment]
@dotnet-bot test Linux tests please

**Comment by leirocks** [comment]
@dotnet-bot test OSX daily_osx_osx_debug_static please

**Comment by leirocks** [comment]
@dotnet-bot test Linux tests please

**Comment by leirocks** [comment]
@dotnet-bot test Linux tests please

**Comment by leirocks** [comment]
@dotnet-bot test Linux tests please

**Comment by leirocks** [comment]
@dotnet-bot test OSX daily_osx_osx_debug_static  please

File: lib/Common/Memory/HeapInfo.h [hunk 7]
```diff
@@ -496,7 +496,16 @@ HeapInfo::RealAlloc(Recycler * recycler, size_t sizeCat, size_t size)
 {
     Assert(HeapInfo::IsAlignedSmallObjectSize(sizeCat));
     auto& bucket = this->GetBucket<(ObjectInfoBits)(attributes & GetBlockTypeBitMask)>(sizeCat);
-    return bucket.template RealAlloc<attributes, nothrow>(recycler, sizeCat, size);
+    if (CONFIG_FLAG(ForceSoftwareWriteBarrier)
```
Comment (digitalinfinity) [inline lib/Common/Memory/HeapInfo.h]: Should this if-clause be in non-Release builds? Otherwise, there is a risk of a perf regression here, no? Same for the other change in this file

File: lib/Common/Memory/LargeHeapBlock.cpp [hunk 8]
```diff
@@ -484,6 +484,9 @@ LargeHeapBlock::AllocFreeListEntry(size_t size, ObjectInfoBits attributes, Large
 
     header->objectIndex = headerIndex;
     header->objectSize = originalSize;
+#ifdef RECYCLER_WRITE_BARRIER
+    header->hasWriteBarrier = (attributes&WithBarrierBit) == WithBarrierBit;
```
Comment (digitalinfinity) [inline lib/Common/Memory/LargeHeapBlock.cpp]: Nit: spacing. Also, do we need to do this explicitly? Can't we tell by the allocator instead? Is there a case where objects are allocated with and without write barrier in the same heap block?

File: lib/Common/Memory/LargeHeapBlock.cpp [hunk 9]
```diff
@@ -970,8 +975,46 @@ LargeHeapBlock::ScanNewImplicitRoots(Recycler * recycler)
 }
 
 #if ENABLE_CONCURRENT_GC
+bool LargeHeapBlock::CheckDirtyOnePage(char* page, RescanFlags flags, bool isWriteBarrier)
```
Comment (digitalinfinity) [inline lib/Common/Memory/LargeHeapBlock.cpp]: >CheckDirtyOnePage [](start = 21, length = 17)

Nit: Prefer the name IsPageDirty.

File: lib/Common/Memory/LargeHeapBlock.cpp [hunk 10]
```diff
@@ -970,8 +975,46 @@ LargeHeapBlock::ScanNewImplicitRoots(Recycler * recycler)
...
+        DWORD pageSize = AutoSystemInfo::PageSize;
+        DWORD const writeWatchFlags = (flags & RescanFlags_ResetWriteWatch ? WRITE_WATCH_FLAG_RESET : 0);
+        void * written = nullptr;
+        UINT ret = GetWriteWatch(writeWatchFlags, page, AutoSystemInfo::PageSize, &written, &count, &pageSize);
+        bool isDirty = (ret != 0) || (count == 1);
+        return isDirty;
+    }
+    else
+    {
+        Assert(UNREACHED);
```
Comment (digitalinfinity) [inline lib/Common/Memory/LargeHeapBlock.cpp]: What is this scenario? Should this be a fail-fast instead of an assert?

File: lib/Common/Memory/Recycler.h [hunk 11]
```diff
@@ -1009,7 +1026,7 @@ class Recycler
 
 #ifdef IDLE_DECOMMIT_ENABLED
     HANDLE concurrentIdleDecommitEvent;
-    DWORD needIdleDecommitSignal;
```
Comment (leirocks) [inline lib/Common/Memory/Recycler.h]: looks the pal only have LONG version of InterlockedCompareExchange, and there was a compiler error, or need to cast between int32 and long

---
In reply to: [88136723](https://github.com/Microsoft/ChakraCore/pull/1968#discussion_r88136723) [](ancestors = 88136723)

File: lib/Common/Memory/Recycler.cpp [hunk 12]
```diff
@@ -4806,16 +4864,25 @@ Recycler::StartBackgroundMark(bool foregroundResetMark, bool foregroundFindRoots
     bool doBackgroundFindRoots = true;
     if (foregroundResetMark || foregroundFindRoots)
     {
-        RECYCLER_PROFILE_EXEC_BEGIN(this, Js::ResetWriteWatchPhase);
-        bool hasWriteWatch = (recyclerPageAllocator.ResetWriteWatch() && recyclerLargeBlockPageAllocator.ResetWriteWatch());
-        RECYCLER_PROFILE_EXEC_END(this, Js::ResetWriteWatchPhase);
-
-        if (!hasWriteWatch)
+        // REVIEW: SWB, if there's only write barrier page change, we don't scan and mark?
```
Comment (digitalinfinity) [inline lib/Common/Memory/Recycler.cpp]: I don't understand this comment

File: lib/Common/Memory/Recycler.cpp [hunk 13]
```diff
@@ -4806,16 +4864,25 @@ Recycler::StartBackgroundMark(bool foregroundResetMark, bool foregroundFindRoots
...
+#if ENABLE_WRITE_WATCH
+        if (!CONFIG_FLAG(ForceSoftwareWriteBarrier))
         {
-            // Disable concurrent mark
-            this->enableConcurrentMark = false;
-            return false;
+            if (!recyclerPageAllocator.ResetWriteWatch() || !recyclerLargeBlockPageAllocator.ResetWriteWatch())
```
Comment (digitalinfinity) [inline lib/Common/Memory/Recycler.cpp]: I'm confused by this- you call ResetWriteWatch here- and then you call it again two lines later. Is that intentional?

File: lib/Common/Memory/PageAllocator.h [hunk 14]
```diff
@@ -479,24 +479,131 @@ class PageAllocatorBase: public PageAllocatorBaseCommon
     static size_t GetAndResetMaxUsedBytes();
 
 #if ENABLE_BACKGROUND_PAGE_FREEING
+    struct FreePageEntry
```
Comment (digitalinfinity) [inline lib/Common/Memory/PageAllocator.h]: Can you add a TODO item to implement a platform agnostic version of interlocked linked lists?

**Comment by digitalinfinity** [comment]
:shipit:

File: lib/Common/Memory/Recycler.cpp (hunk 13 above)
Comment (leirocks) [inline lib/Common/Memory/Recycler.cpp]: that's a mistake. I was thinking why didn't we reset barrier here and trying fix it.  after that I changed to just leave a REVIEW comment here but didn't remove the above line. let me remove it

---
In reply to: [88137857](https://github.com/Microsoft/ChakraCore/pull/1968#discussion_r88137857) [](ancestors = 88137857)

File: lib/Common/Memory/Recycler.cpp (hunk 12 above)
Comment (leirocks) [inline lib/Common/Memory/Recycler.cpp]: see comment bellow

---
In reply to: [88137787](https://github.com/Microsoft/ChakraCore/pull/1968#discussion_r88137787) [](ancestors = 88137787)

File: lib/Common/Memory/Recycler.h [hunk 15]
```diff
@@ -1619,8 +1654,10 @@ class Recycler
     size_t TryMarkArenaMemoryBlockList(ArenaMemoryBlock * memoryBlocks);
     size_t TryMarkBigBlockList(BigBlock * memoryBlocks);
 #if ENABLE_CONCURRENT_GC
+#if FALSE // REVIEW: remove this code since not using
     size_t TryMarkBigBlockListWithWriteWatch(BigBlock * memoryBlocks);
```
Comment (leirocks) [inline lib/Common/Memory/Recycler.h]: the caller is comment out with #if FALSE, and saying 'Re-enable this in the future.' I'm just removing the definition. let me add a TODO on the caller and investigate in the future

---
In reply to: [88136880](https://github.com/Microsoft/ChakraCore/pull/1968#discussion_r88136880) [](ancestors = 88136880)

File: lib/Common/Memory/LargeHeapBlock.cpp (hunk 10 above)
Comment (leirocks) [inline lib/Common/Memory/LargeHeapBlock.cpp]: in case we disabled hardware write barrier, but we are still checking hardware write barrier on the page. yes it's better to be a fail fast, let me change it


---
In reply to: [88135185](https://github.com/Microsoft/ChakraCore/pull/1968#discussion_r88135185) [](ancestors = 88135185)

File: lib/Common/Memory/LargeHeapBlock.cpp (hunk 8 above)
Comment (leirocks) [inline lib/Common/Memory/LargeHeapBlock.cpp]: this is told by the allocator, passing in the WithBarrierBit. currently the mix mode won't happen because we either turn on all with writebarrier or, only turn on for function body, which is not in large block. in the future if we selectively enable more write barrier allocation on windows, we will hit such scenario. and I had a comment in the code to use dedicated page allocator for large block with barrier.
fixing the spaces

---
In reply to: [88134866](https://github.com/Microsoft/ChakraCore/pull/1968#discussion_r88134866) [](ancestors = 88134866)

File: lib/Common/Memory/HeapInfo.h (hunk 7 above)
Comment (leirocks) [inline lib/Common/Memory/HeapInfo.h]: my plan was removing this flag after the feature is done. I think yes there's perf regression but not sure how much will it be.
let put it under GLOBAL_ENABLE_WRITE_BARRIER, so only special build will have this on windows

---
In reply to: [88134740](https://github.com/Microsoft/ChakraCore/pull/1968#discussion_r88134740) [](ancestors = 88134740)

**Comment by leirocks** [comment]
@dotnet-bot test Linux tests please

**Comment by leirocks** [comment]
@dotnet-bot test OSX daily_osx_osx_debug_static please
